### ISSUE TITLE
Use of uninitialized value in division (/) at ./scripts/multi-bleu.perl line 139, <STDIN> line 1553.

### ISSUE THREAD

**Issue by fantasysee** [tnq177/transformers_without_tears#4]
Dear Toan Nguyen,

Thank you so much for your open-source transformer repo.

I am trying to modify some of the key modules with the quantization function. However, sometimes the error message is reported:

> Use of uninitialized value in division (/) at ./scripts/multi-bleu.perl line 139, <STDIN> line 1553."

, and then the BLEU metric would be 0. And then the training process failed to coverage, and the BLEU would maintain 0.

Would you please tell me what may be the reason triggering this error, and how to fix it?

Thanks in advance.

**Comment by tnq177** [comment]
hi @fantasysee. I've never used quantization before so I honestly don't know. But I think those modifications causes training to diverge and *therefore* the translation is gibberish which means BLEU = 0 which triggers that the error you see. So you probably have to check your changes again  😬 . Please note that depending on your dataset, early in the training process if the model is not good enough, translation will have BLEU = 0 and you still see that error message but it will go away as model is well trained. Please make sure to use the same version of python/pytorch as mentioned in readme. I've never tested with newer versions.

**Comment by tnq177** [comment]
Just curious about what you're trying to do though? Are you trying to train with 8-bit or are you trying to quantize a pretrained model? I might look into the later case cuz i'm curious about it too - but i can't promise anything for now. For the former, I think it's tricky and I suggest you should try other code that supports it already. I'm not sure which one does, but looks like fairseq has a section about quantization https://github.com/pytorch/fairseq.

**Comment by fantasysee** [comment]
> hi @fantasysee. I've never used quantization before so I honestly don't know. But I think those modifications causes training to diverge and _therefore_ the translation is gibberish which means BLEU = 0 which triggers that the error you see. So you probably have to check your changes again 😬 . Please note that depending on your dataset, early in the training process if the model is not good enough, translation will have BLEU = 0 and you still see that error message but it will go away as model is well trained. Please make sure to use the same version of python/pytorch as mentioned in readme. I've never tested with newer versions.

Hi @tnq177 , thank you so much for your warm reply!

I may miss something important to describe the phenomenon: currently, I just quantized all the softmax input in the attention module with 8-bit integer; and the quantization strategy is deployed from scratch. The training process seems good at first since the metrics, `smppl` and `ppl`, dropped as the same as the FP version, and `BLEU` increased. NEVERTHELESS, the `BLEU` achieved 25.83 at epoch 28, and then sharply dropped to 0.41 at epoch 29, with the error message:

> Use of uninitialized value in division (/) at ./scripts/multi-bleu.perl line 139, line 1553.

I am not sure if it is a common phenomenon. I think maybe find out what the error message means may help me fix this problem.

Later I would like to explore the possibility to train a transformer with 8-bit.

Thanks again for your advice!!!

**Comment by fantasysee** [comment]
@tnq177 You're right!!! I think this [example](https://github.com/pytorch/fairseq/blob/master/examples/quant_noise/README.md) in fairseq could satisfy my need. Thanks for your clue!!!
